### ISSUE TITLE
Batch 5: End-to-end 10K MERFISH evaluation, timing, and report

### ISSUE THREAD

**Issue by Trecek** [TalonT-Org/spectral-init#203]
## Summary

Run the full MERFISH 10K evaluation pipeline end-to-end with all 17 metrics,
add timing and memory instrumentation, and produce the benchmark report. This
is Batch 5 from `docs/merfish-umap-benchmark-plan.md` — the capstone that
produces the first complete Rust-vs-Python comparison on real spatial
transcriptomics data.

**This is a manual live-session task**, not a CI test or automated pipeline.

## Prerequisites

All of these must be merged before starting:

- [x] #190 — 10K subset committed
- [x] #192 — `generate_merfish_comparisons.py` pipeline
- [ ] #195 — Category B spatial metrics
- [ ] #197 — Category C cluster metrics
- [ ] #199 — Category D global metrics
- [ ] #200 — Wire B/C/D metrics into pipeline

### 5.1 — End-to-End Pipeline Run

Run `./tests/visual_eval/run_merfish_eval.sh` with all 17 metrics wired in.
Verify:

- [ ] Phase 1 (Python baseline) completes without errors
- [ ] Phase 2 (Rust `spectral_init` via `cargo nextest --profile merfish-eval`)
      completes and produces `merfish_10k_rust_init.npy`
- [ ] Phase 3 (three-way comparison) completes and produces
      `merfish_10k_metrics.json` with all Category A/B/C/D metrics
- [ ] All plots generate: `merfish_10k_baseline.png`,
      `merfish_10k_comparison.png`, `merfish_10k_overlay.png`,
      `merfish_10k_three_way_overlay.png`
- [ ] PASS/FAIL summary prints correctly with the SNA gate included

**Expected result pattern:** Trustworthiness PASS, Silhouette PASS, SNA
PASS (or close), Procrustes FAIL, Pairwise Corr FAIL. This is the same
"quality PASS, geometry FAIL" pattern seen across all Tier 1/2 datasets
and confirmed in the initial 10K run.

### 5.2 — Timing and Memory Instrumentation

Add per-phase timing and peak memory measurement to
`generate_merfish_comparisons.py`:

**Timing (wall-clock, per phase):**

| Phase | What to Measure |
|---|---|
| Data loading | Time to load `.npz` files from `merfish_data/` |
| Preprocessing | `normalize_total` + `log1p` + `scale` + `PCA` + `neighbors` |
| Python spectral init | `spectral_layout()` call only |
| Python SGD | `UMAP.fit_transform()` minus spectral init |
| Graph export | `export_graph()` write time |
| Rust spectral init | `cargo nextest` wall time (from shell script or `time` wrapper) |
| Rust-init SGD | `UMAP(init=rust_init).fit_transform()` |
| Random SGD | `UMAP(init='random').fit_transform()` |
| Metric computation | All `compute_*_metrics()` calls |
| Plot generation | All `_make_*_plot()` calls |

**Memory (peak RSS):**
```python
import resource
peak_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
```
Record before and after each major phase. For Rust, use
`/usr/bin/time -v cargo nextest ...` and parse `Maximum resident set size`.

**Output:**
- `tests/visual_eval/output/merfish_10k_timing.json`:
  ```json
  {
    "data_loading_s": 0.3,
    "preprocessing_s": 2.1,
    "python_spectral_init_s": 0.8,
    "python_sgd_s": 12.4,
    "graph_export_s": 0.1,
    "rust_spectral_init_s": 1.0,
    "rust_init_sgd_s": 11.9,
    "random_sgd_s": 12.2,
    "metrics_s": 3.5,
    "plots_s": 1.8,
    "total_baseline_s": 29.8,
    "total_compare_s": 37.0
  }
  ```
- `tests/visual_eval/output/merfish_10k_memory.json`:
  ```json
  {
    "peak_rss_baseline_mb": 450,
    "peak_rss_compare_mb": 520,
    "rust_peak_rss_mb": 85
  }
  ```

### 5.3 — Benchmark Report

Compile all results into `docs/merfish-10k-benchmark-report.md`:

**Section 1: Dataset Summary**
- MERFISH Zhuang-ABCA-1, 10K spatially-stratified subset
- 1,122 genes, N sections sampled, cell-type distribution
- Preprocessing config (winner from #193 sweep, or defaults)

**Section 2: Quality Results (All 17 Metrics)**

Full table comparing Python spectral init, Rust spectral init, and Random init:

| Metric | Category | Python | Rust | Random | Rust PASS/FAIL |
|---|---|---|---|---|---|
| Trustworthiness | A | ... | ... | ... | ... |
| Silhouette | A | ... | ... | ... | ... |
| Procrustes | A | ... | ... | ... | ... |
| Pairwise Corr | A | ... | ... | ... | ... |
| SNA | B | ... | ... | ... | ... |
| Spatial Dist Corr | B | ... | ... | ... | ... |
| Moran's I (max) | B | ... | ... | ... | ... |
| CHAOS | B | ... | ... | ... | ... |
| PAS | B | ... | ... | ... | ... |
| ARI | C | ... | ... | ... | ... |
| NMI | C | ... | ... | ... | ... |
| Cell-type Purity | C | ... | ... | ... | ... |
| Triplet Accuracy | D | ... | ... | ... | ... |
| Shepard (Spearman) | D | ... | ... | ... | ... |
| Centroid Dist Corr | D | ... | ... | ... | ... |
| kNN Preservation | D | ... | ... | ... | ... |

**Section 3: Timing**
- Per-phase wall-clock breakdown table
- Key comparison: Python spectral init time vs Rust spectral init time
- Note: at 10K cells both are sub-second; scaling advantage appears at 100K+

**Section 4: Memory**
- Peak RSS for Python baseline, Python compare, Rust export

**Section 5: Plots**
- Embed or reference all generated `.png` files
- `merfish_10k_baseline.png` — 2×2 grid
- `merfish_10k_comparison.png` — 2×3 three-way comparison
- `merfish_10k_overlay.png` — Python vs Rust overlay
- `merfish_10k_three_way_overlay.png` — all three overlaid

**Section 6: Interpretation**
- Does Rust spectral init produce equivalent downstream embeddings?
  (Answer from Category A metrics)
- Does Rust init preserve spatial tissue organization?
  (Answer from Category B metrics — the novel contribution)
- Does Rust init preserve cell-type cluster structure?
  (Answer from Category C metrics)
- How does Rust compare on global structure?
  (Answer from Category D metrics)
- Is Random init a useful lower bound? (Random should score worse on
  everything — if not, the dataset may be "too easy")

**Section 7: Conclusions and Next Steps**
- Summary verdict on Rust spectral init quality at 10K
- Whether to proceed to 100K scaling study
- Any preprocessing adjustments indicated by the sweep results
- Known limitations (single biological sample, 10K subset only)

## How to Run

```bash
# Full pipeline (after all metric PRs are merged)
cd /home/talon/projects/spectral-init
micromamba activate spectral-test
./tests/visual_eval/run_merfish_eval.sh

# Or phase by phase:
python tests/visual_eval/generate_merfish_comparisons.py --phase baseline
cargo nextest run --profile merfish-eval --run-ignored all --features testing
python tests/visual_eval/generate_merfish_comparisons.py --phase compare

# View results:
cat tests/visual_eval/output/merfish_10k_metrics.json | python -m json.tool
cat tests/visual_eval/output/merfish_10k_timing.json | python -m json.tool
```

## Evaluation Checklist

After running the full pipeline, manually verify:

- [ ] `merfish_10k_metrics.json` contains all 4 metric categories
- [ ] Trustworthiness and Silhouette deltas are within PASS thresholds
- [ ] Spatial metrics (SNA, Moran's I) are computed and non-NaN
- [ ] Random init scores worse than both Python and Rust on spatial metrics
      (if not, flag the dataset as "too easy" for spatial evaluation)
- [ ] Plots are visually reasonable (cells colored by type show structure)
- [ ] Timing JSON has all phase breakdowns
- [ ] The report accurately reflects the numbers
### RUN — Pipeline Execution

- **REQ-RUN-001:** The full three-phase pipeline (`run_merfish_eval.sh`) must complete without errors when all Batch 4 PRs are merged.
- **REQ-RUN-002:** `merfish_10k_metrics.json` must contain metric values for all 4 categories (A, B, C, D) under `python_spectral`, `rust_spectral`, and `random` keys.
- **REQ-RUN-003:** All 4 plot files must be generated in `tests/visual_eval/output/`.

### TIME — Timing Instrumentation

- **REQ-TIME-001:** Wall-clock time must be recorded separately for each major phase: data loading, preprocessing, Python spectral init, Python SGD, graph export, Rust spectral init, Rust-init SGD, random SGD, metrics computation, and plot generation.
- **REQ-TIME-002:** Timing data must be saved as `merfish_10k_timing.json` in `tests/visual_eval/output/`.
- **REQ-TIME-003:** Python spectral init time must be measured by timing the `spectral_layout()` call in isolation, not including kNN construction or SGD.

### MEM — Memory Instrumentation

- **REQ-MEM-001:** Peak RSS must be recorded via `resource.getrusage(RUSAGE_SELF).ru_maxrss` for Python phases.
- **REQ-MEM-002:** Rust peak RSS must be captured via `/usr/bin/time -v` or equivalent.
- **REQ-MEM-003:** Memory data must be saved as `merfish_10k_memory.json` in `tests/visual_eval/output/`.

### REPORT — Benchmark Report

- **REQ-REPORT-001:** A report file `docs/merfish-10k-benchmark-report.md` must be created containing dataset summary, full 17-metric comparison table, timing breakdown, memory profile, plot references, interpretation, and conclusions.
- **REQ-REPORT-002:** The report must include the actual numeric values from the pipeline run, not placeholder text.
- **REQ-REPORT-003:** The report must state a clear verdict on whether Rust spectral init is equivalent to Python on the MERFISH 10K subset.
- **REQ-REPORT-004:** The report must identify next steps (proceed to 100K scaling, adjust preprocessing, etc.).

### VALID — Evaluation Criteria

- **REQ-VALID-001:** Random init must score worse than both Python and Rust spectral init on spatial metrics (SNA, Moran's I). If it does not, the report must flag the dataset as potentially "too easy" for spatial evaluation.
- **REQ-VALID-002:** The expected pattern is "quality PASS (trustworthiness, silhouette, SNA), geometry FAIL (Procrustes, pairwise corr)" — deviations from this pattern must be discussed in the report.
### RUN — Pipeline Execution

- **REQ-RUN-001:** The full three-phase pipeline (`run_merfish_eval.sh`) must complete without errors when all Batch 4 PRs are merged.
- **REQ-RUN-002:** `merfish_10k_metrics.json` must contain metric values for all 4 categories (A, B, C, D) under `python_spectral`, `rust_spectral`, and `random` keys.
- **REQ-RUN-003:** All 4 plot files must be generated in `tests/visual_eval/output/`.

### TIME — Timing Instrumentation

- **REQ-TIME-001:** Wall-clock time must be recorded separately for each major phase: data loading, preprocessing, Python spectral init, Python SGD, graph export, Rust spectral init, Rust-init SGD, random SGD, metrics computation, and plot generation.
- **REQ-TIME-002:** Timing data must be saved as `merfish_10k_timing.json` in `tests/visual_eval/output/`.
- **REQ-TIME-003:** Python spectral init time must be measured by timing the `spectral_layout()` call in isolation, not including kNN construction or SGD.

### MEM — Memory Instrumentation

- **REQ-MEM-001:** Peak RSS must be recorded via `resource.getrusage(RUSAGE_SELF).ru_maxrss` for Python phases.
- **REQ-MEM-002:** Rust peak RSS must be captured via `/usr/bin/time -v` or equivalent.
- **REQ-MEM-003:** Memory data must be saved as `merfish_10k_memory.json` in `tests/visual_eval/output/`.

### REPORT — Benchmark Report

- **REQ-REPORT-001:** A report file `docs/merfish-10k-benchmark-report.md` must be created containing dataset summary, full 17-metric comparison table, timing breakdown, memory profile, plot references, interpretation, and conclusions.
- **REQ-REPORT-002:** The report must include the actual numeric values from the pipeline run, not placeholder text.
- **REQ-REPORT-003:** The report must state a clear verdict on whether Rust spectral init is equivalent to Python on the MERFISH 10K subset.
- **REQ-REPORT-004:** The report must identify next steps (proceed to 100K scaling, adjust preprocessing, etc.).

### VALID — Evaluation Criteria

- **REQ-VALID-001:** Random init must score worse than both Python and Rust spectral init on spatial metrics (SNA, Moran's I). If it does not, the report must flag the dataset as potentially "too easy" for spatial evaluation.
- **REQ-VALID-002:** The expected pattern is "quality PASS (trustworthiness, silhouette, SNA), geometry FAIL (Procrustes, pairwise corr)" — deviations from this pattern must be discussed in the report.

## Execution via Research Recipe

This issue can be executed through the local `.autoskillit/recipes/research.yaml` recipe. **Run Phase 1 first** to scope and produce the experiment design for human review:

```
task: "End-to-end 10K MERFISH evaluation: timing instrumentation, all 17 metrics, benchmark report"
source_dir: /home/talon/projects/spectral-init
base_branch: main
setup_phases: true
```

Phase 1 produces an experiment plan issue. After human review and approval, resume with `issue_number` set to run Phase 2 (implement, run experiment, write report, open PR).
